Sources/URI/Parser/Parser.swift

```diff
@@ -1,119 +1,115 @@
 import Core
 import Transport
-
-public final class URIParser: StaticDataBuffer {
-
-    public enum Error: Swift.Error {
-        case invalidPercentEncoding
-        case unsupportedURICharacter(Byte)
-    }
-
-    // If we have authority, we should also have scheme?
-    let existingHost: Bytes?
-
-    /**
-        The most common form of Request-URI is that used to identify a
-        resource on an origin server or gateway. In this case the absolute
-        path of the URI MUST be transmitted (see section 3.2.1, abs_path) as
-        the Request-URI, and the network location of the URI (authority) MUST
-        be transmitted in a Host header field. For example, a client wishing
-        to retrieve the resource above directly from the origin server would
-        create a TCP connection to port 80 of the host "www.w3.org" and send
-        the lines:
-
-        GET /pub/WWW/TheProject.html HTTP/1.1
-        Host: www.w3.org
-
-        If host exists, and scheme exists, use those
-    */
-    public init(bytes: Bytes, existingHost: String? = nil) {
-        self.existingHost = existingHost?.makeBytes()
-        super.init(bytes: bytes)
-    }
-
-    // MARK: Paser URI
-
-    /**
-        Main parsing function
-    */
-    internal func parse() throws -> URI {
-        let (schemeBytes, authorityBytes, pathBytes, queryBytes, fragmentBytes) = try parseComponents()
-        let (usernameBytes, infoBytes, hostBytes, portBytes) = try parse(authority: authorityBytes)
-
-        /*
-            ***** [WARNING] *****
-
-            do NOT attempt to percent decode before THIS point
-        */
-        let scheme = try schemeBytes.percentDecodedString()
-        let username = try usernameBytes?.percentDecodedString()
-        let info = try infoBytes?.percentDecodedString()
-
-        let userInfo: URI.UserInfo?
-        if let username = username, !username.isEmpty {
-            userInfo = URI.UserInfo(
-                username: username,
-                info: info
-            )
+import CHTTP
+
+/// Parses URIs from bytes.
+public final class URIParser {
+    /// Use a shared parser since URI parser doesn't
+    /// require special configuration
+    public static let shared = URIParser()
+    
+    /// Creates a new URI parser.
+    public init() {}
+    
+    /// Parses a URI from the supplied bytes.
+    public func parse(bytes: Bytes) -> URI {
+        // create url results struct
+        var url = http_parser_url()
+        http_parser_url_init(&url)
+        
+        // parse url
+        http_parser_parse_url(bytes.makeCBytes(), bytes.count, 0, &url)
+        
+        // fetch offsets from result
+        let (scheme, hostname, port, path, query, fragment, userinfo) = url.field_data
+        
+        // parse uri info
+        let info: URI.UserInfo?
+        if userinfo.len > 0, let bytes = bytes.bytes(for: userinfo) {
+            let parts = bytes.split(separator: .colon, maxSplits: 1)
```

Comment:
I'm not positive this allows empty subsequences by default or not, we should maybe be explicit `allowsEmptySubsequences: true`